### PR TITLE
HTTPCLIENT-2082

### PR DESCRIPTION
**Pull request by rafiek**
Adding the connection socket lookup as a parameter for the builder makes it more flexible to use this builder.

### CODE REVIEW

File: httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManagerBuilder.java [hunk 1]
```diff
@@ -70,10 +74,13 @@
  */
 public class PoolingHttpClientConnectionManagerBuilder {
 
+    private Lookup<ConnectionSocketFactory> connectionSocketFactoryRegistry;
     private HttpConnectionFactory<ManagedHttpClientConnection> connectionFactory;
     private LayeredConnectionSocketFactory sslSocketFactory;
     private SchemePortResolver schemePortResolver;
     private DnsResolver dnsResolver;
+
+    private StrictConnPool<HttpRoute, ManagedHttpClientConnection> managedConnPool;
```
Comment (ok2c) [inline httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManagerBuilder.java:83]: @rafiek Why do we need this variable? It invalidates `PoolConcurrencyPolicy` parameter.

File: httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManagerBuilder.java [hunk 2]
```diff
@@ -70,10 +74,13 @@
  */
 public class PoolingHttpClientConnectionManagerBuilder {
 
+    private Lookup<ConnectionSocketFactory> connectionSocketFactoryRegistry;
```
Comment (ok2c) [inline httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManagerBuilder.java:77]: @rafiek Same problem. This variable invalidates `sslSocketFactory`. What does it get us?

File: httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManagerBuilder.java (hunk 1 above)
Comment (rafiek) [reply httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManagerBuilder.java:83]: It is used in tests, so I added the option to make it part of the `build()`. When omitted, you can still use `PoolConcurrencyPolicy`.

File: httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManagerBuilder.java (hunk 2 above)
Comment (rafiek) [reply httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManagerBuilder.java:77]: When omitting the use of this variable for `'build()`, you can still use `sslSocketFactory` as part of `build()`.

**Comment by ok2c** [comment]
@rafiek I understand that. However we have a very similar situation with `HttpClientBuilder` in HC 4.x and it has been causing quite a bit of confusion and troubles with users who never get around to reading the API documentation. I do not want to repeat the same situation in HC 5.x.

**Comment by rafiek** [comment]
> @rafiek I understand that. However we have a very similar situation with `HttpClientBuilder` in HC 4.x and it has been causing quite a bit of confusion and troubles with users who never get around to reading the API documentation. I do not want to repeat the same situation in HC 5.x.

Aah, did not know that. Then how about this. I was preparing this feature to make it easier to implement a programmatic SOCKS based HttpClient. The `HttpClientBuilder` uses the `PoolingHttpClientConnectionManagerBuilder` to build a default HttpClientConnectionManager, but doesn't take into consideration to use `systemProperties`. When it does, then it would be very convenient that the builder takes in `connectionSocketFactoryRegistry` as a step for `build()`. This was part of [HTTPCLIENT-1966](https://issues.apache.org/jira/browse/HTTPCLIENT-1966), but I closed the pull request because it was preferred to implement it in HC 5.x.

Should I also add the changes for HTTPCLIENT-1966 in this pull request. Making the scope bigger and the intention for these changes more clear. Maybe adding more javadoc also?

**Comment by ok2c** [comment]
@rafiek I initially thought you were going to add support for SOCKS related system properties but with this PR you have veered into an unexpected direction.

**Comment by rafiek** [comment]
> @rafiek I initially thought you were going to add support for SOCKS related system properties but with this PR you have veered into an unexpected direction.

No worries. Hang tight. I'll steer this boat back into familiar waters. I'll contact you in Jira when I've updated this pull request.

**Pull request closed without merging**